pkg/gds/admin/v2/csurf_test.go

```diff
@@ -0,0 +1,175 @@
+package admin_test
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/require"
+	"github.com/trisacrypto/directory/pkg/gds/admin/v2"
+)
+
+func TestDoubleCookies(t *testing.T) {
+	// Test both the DoubleCookie middleware and the SetDoubleCookieTokens handler.
+	router := gin.New()
+
+	// Add login route that sets the cookies
+	router.GET("/login", func(c *gin.Context) {
+		err := admin.SetDoubleCookieTokens(c, time.Now().Add(time.Minute*10).Unix())
+		require.NoError(t, err)
+		c.JSON(http.StatusOK, gin.H{"success": true})
+	})
+
+	// Add request route that requires double cookie submit
+	router.POST("/action", admin.DoubleCookie(), func(c *gin.Context) {
+		c.JSON(http.StatusCreated, gin.H{"success": true})
+	})
+
+	// Create an http client with a cookie jar
+	jar, err := cookiejar.New(nil)
+	require.NoError(t, err)
+	client := http.Client{
+		Jar: jar,
+	}
+
+	// Create the test server with the CSRF protected router
+	server := httptest.NewServer(router)
+
+	// Attempt to make a request to the server that is not CSRF protected
+	req, err := http.NewRequest(http.MethodPost, server.URL+"/action", nil)
+	require.NoError(t, err)
+
+	// Ensure the request is Forbidden
+	rep, err := client.Do(req)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusForbidden, rep.StatusCode)
+
+	// Check the data in the response
+	data, err := readJSON(rep)
+	require.NoError(t, err)
+	require.Contains(t, data, "error")
+	require.Contains(t, data, "success")
```

Comment:
Can you help me understand why it's an error and a success here?

Reply:
The json response looks like this:

```json
{
    "success": false, 
    "error": "csrf token validation failed"
}
```

So here I'm testing that the data contains the "success" key and then on line 59 that it is false. 

The `success` bool is very much a throwback to some old school Ben development, which I would be happy to chat about during a happy hour with pleasant cocktails :-)